common/src/sphere_collider.rs

```diff
@@ -0,0 +1,294 @@
+    let mip_pos_a = math::mip(&ray.position, line_position);
+    let mip_dir_a = math::mip(&ray.direction, line_position);
+    let mip_pos_b = math::mip(&ray.position, line_direction);
+    let mip_dir_b = math::mip(&ray.direction, line_direction);
+
+    solve_quadratic(
+        mip_pos_a.powi(2) - mip_pos_b.powi(2) - cosh_radius.powi(2),
+        mip_pos_a * mip_dir_a - mip_pos_b * mip_dir_b,
+        mip_dir_a.powi(2) - mip_dir_b.powi(2) + cosh_radius.powi(2),
+    )
+}
+
+/// Finds the tanh of the distance a sphere will have to travel along a ray before it
+/// intersects the given plane.
+///
+/// Returns NaN if there's no such intersection
+fn solve_sphere_plane_intersection(
+    ray: &Ray,
+    plane_normal: &na::Vector4<f32>,
+    sinh_radius: f32,
+) -> f32 {
+    let mip_pos_a = math::mip(&ray.position, plane_normal);
+    let mip_dir_a = math::mip(&ray.direction, plane_normal);
+
+    solve_quadratic(
+        mip_pos_a.powi(2) - sinh_radius.powi(2),
+        mip_pos_a * mip_dir_a,
+        mip_dir_a.powi(2) + sinh_radius.powi(2),
+    )
+}
+
+/// Finds the lower solution `x` of `constant_term + 2 * half_linear_term * x + quadratic_term * x * x == 0`
+///
+/// Returns NaN if no such solution exists.
+///
+/// If a small perturbation to these terms would result in a solution of `x == 0.0`, this function has logic to
+/// to return 0.0 if three conditions hold in the context of collision checking:
+/// 1. The collider must be intersecting the object. This manifests as `constant_term <= 0.0`.
+/// 2. The collider must not be too far inside the object. This manifests as `constant_term >= -EPSILON`.
+/// 3. The direction of motion must be towards the collider. This manifests as `double_linear_term < 0.0`.
+fn solve_quadratic(constant_term: f32, half_linear_term: f32, quadratic_term: f32) -> f32 {
+    const EPSILON: f32 = 1e-4;
+
+    // Extra logic to ensure precision issues don't allow a collider to clip through a surface
+    if (-EPSILON..=0.0).contains(&constant_term) && half_linear_term < 0.0 {
+        return 0.0;
+    }
+
+    let discriminant = half_linear_term * half_linear_term - quadratic_term * constant_term;
+
+    // We use an alternative quadratic formula to ensure that we return a positive number if `constant_term > 0.0`.
+    // Otherwise, the edge case of a small positive `constant_term` could be mishandled.
+    // Note that discriminant can be negative, which allows this function to return NaN when there is no solution.
+    constant_term / (-half_linear_term + discriminant.sqrt())
+}
+
+/// Converts from t-u-v coordinates to x-y-z coordinates. t-u-v coordinates are a permuted version of x-y-z coordinates.
+/// `t_axis` determines which of the three x-y-z coordinates corresponds to the t-coordinate. This function works with
+/// any indexable entity with at least three entries. Any entry after the third entry is ignored.
+fn tuv_to_xyz<T: std::ops::IndexMut<usize, Output = N>, N: Copy>(t_axis: usize, tuv: T) -> T {
```

Comment:
`fn tuv_to_xyz<T: Copy, const D: usize>(t_axis: usize, tuv: [T; D])`, then calling `.into()` on vectors where needed, might be a little more straightforward. Your call, though.